app/src/main/java/com/UserDatabase/UserDatabase.java

```diff
@@ -0,0 +1,264 @@
+package com.UserDatabase;
+
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import com.google.gson.reflect.TypeToken;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Random;
+import java.util.Set;
+
+import static org.junit.Assert.assertEquals;
+
+public abstract class UserDatabase {
+    protected final String LogTag = "Database";
+
+    // Path of database file
+    private final File Filepath;
+
+    // Database identifier, to ensure safe loading
+    private final String Id;
+    // Vector target length, to validate database inputs
+    private final int VectorLength;
+
+    // protected ArrayList<UserRecord> usersRecords;
+    protected Map<String, float[]> usersRecords;
+
+
+    UserDatabase(File filepath, String id, int vectorLength) {
+        this.Filepath = filepath;
+        this.Id = id;
+        this.VectorLength = vectorLength;
+
+        usersRecords = new HashMap<String, float[]>();
+
+        // Load Database on creation
+        loadDatabase();
+    }
+
+    /**
+     * Find closest record from the database. Algorithm and time complexity is dependent
+     * on database type.
+     *
+     * @param vector of n-dimensions, for which the closest equivalent wil be found.
+     * @return closest UserRecord.
+     */
+    abstract UserRecord findClosestRecord(float[] vector);
+
+    /**
+     * Add new UserRecord to the database.
+     *
+     * @param userRecord to add to database.
+     * @return True if record was added. False if user already exists.
+     */
+    public boolean addUserRecord(UserRecord userRecord) {
+        // Check correctness of vector length
+        if(userRecord.vector.length == VectorLength){
+
+            // Check if user already exists in database
+            if (!usersRecords.containsKey(userRecord.username)) {
+                usersRecords.put(userRecord.username, userRecord.vector);
+
+                // Serialize database immediately
+                // TODO: Later on it might be reasonable to save database on application closure (faster)
+                saveDatabase();
+
+                return true;
+            }

```

Comment:
there should be some assertation on case, when user name is already defined.

Reply:
There should not be one. In such case the returned value is false.

Reply:
+1